services/supervisor/tests/supervisor/migrations/migration_service.rs

```diff
@@ -0,0 +1,410 @@
+        }
+    }
+
+    pub(crate) fn verify_schema(snapshot: Prefixed<'_, &dyn Snapshot>) {
+        let users = super::USERS;
+
+        let schema = Schema::new(snapshot.clone());
+        for user in users {
+            let (key, _) = user.keypair();
+            let wallet = schema
+                .wallets
+                .get(&key)
+                .expect("V05: User wallet not found");
+            assert_eq!(wallet.balance, user.balance);
+            assert_eq!(wallet.first_name, user.first_name);
+            assert_eq!(wallet.last_name, user.last_name);
+        }
+        assert_eq!(schema.wallets.iter().count(), users.len());
+
+        let summary = schema.summary.get().unwrap();
+        assert_eq!(summary.ticker, super::SERVICE_NAME);
+        assert_eq!(
+            summary.total_balance,
+            users.iter().map(|user| user.balance).sum::<u64>()
+        );
+
+        // Check that the outdated index has been deleted.
+        assert_eq!(snapshot.index_type("total_balance"), None);
+    }
+}
+
+/// First migration script. Merkelizes the wallets table and records the total number of tokens.
+fn merkelize_wallets(ctx: &mut MigrationContext) -> Result<(), MigrationError> {
+    let old_schema = v01::Schema::new(ctx.helper.old_data());
+    let mut new_schema = v02::Schema::new(ctx.helper.new_data());
+
+    let mut total_balance = 0;
+    for (key, wallet) in &old_schema.wallets {
+        total_balance += wallet.balance;
+        new_schema.wallets.put(&key, wallet);
+    }
+    new_schema.total_balance.set(total_balance);
+    Ok(())
+}
+
+/// Second migration script. Transforms the wallet type and reorganizes the service summary.
+fn transform_wallet_type(ctx: &mut MigrationContext) -> Result<(), MigrationError> {
+    let old_schema = v02::Schema::new(ctx.helper.old_data());
+    let mut new_schema = v05::Schema::new(ctx.helper.new_data());
+
+    let total_balance = old_schema.total_balance.get().unwrap_or(0);
+    new_schema.summary.set(v05::Summary {
+        ticker: ctx.instance_spec.name.clone(),
+        total_balance,
+    });
+    ctx.helper.new_data().create_tombstone("total_balance");
+
+    for (key, wallet) in &old_schema.wallets {
+        let name_parts: Vec<_> = wallet.username.rsplitn(2, ' ').collect();
+        let (first_name, last_name) = match &name_parts[..] {
+            [first_name] => (*first_name, ""),
+            [last_name, first_name] => (*first_name, *last_name),
+            _ => unreachable!(),
+        };
+        let new_wallet = v05::Wallet {
+            first_name: first_name.to_owned(),
+            last_name: last_name.to_owned(),
+            balance: wallet.balance,
+        };
+        new_schema.wallets.put(&key, new_wallet);
+    }
+    Ok(())
+}
+
+/// Third migration script. Always fails.
+fn failing_migration(_ctx: &mut MigrationContext) -> Result<(), MigrationError> {
+    Err(MigrationError::Custom("This migration always fails".into()))
```

Comment:
Nit: There is `MigrationError::new` method, which is more idiomatic in this case.